course5/homework10/llist.h

```diff
@@ -0,0 +1,184 @@
+#pragma once
+#include <memory>
+#include <utility>
+#include <iostream>
+
+namespace course 
+{
+template<class T>
+class llist
+{
+private:
+        struct node
+        {
+                node() { }
+                T value;
+                std::unique_ptr<node> next;
+        };
+        std::unique_ptr<node> first;
+        int len = 0;
+public:
+ /*       struct iterator
+        {
+            iterator() : node_ptr(nullptr) { }
+            iterator(node* ptr) : node_ptr(ptr) {}
+            iterator(const iterator& other) : node_ptr(other.node_ptr) { }
+            iterator(iterator&& other) : node_ptr(other.node_ptr) { }
+            iterator& operator=(const iterator& other) { return other.node_ptr; }
+            iterator& operator=(iterator&& other) { return other.node_ptr; }
+            bool operator==(const iterator& other) { return node_ptr == other.node_ptr; }
+            bool operator!=(const iterator& other) { return node_ptr != other.node_ptr; }
+            iterator& operator++() 
+            {
+                if (node_ptr != nullptr)
+                {
+                    node_ptr = node_ptr->next.get();
+                    return *this;
+                }
+                else
+                {
+                    throw std::range_error("End of list");
+                }
+            }
+            T& operator*() { return node_ptr->value; }
+            node* node_ptr;
+        };*/
+        template<class P, class Q>
```

Comment:
This class is already a template by the virtual of being nested under one (the fully-qualified name of the class is ```course::llist<T>::iterator_class```. There is really no need to make it even more templaty.